### PR TITLE
bpo-42064: Pass module state to `sqlite3` UDF callbacks

### PR DESCRIPTION
**Pull request by erlend-aasland**
- Establish common callback context struct
- Convert UDF callbacks to fetch module state from callback context

https://bugs.python.org/issue42064

Automerge-Triggered-By: GH:encukou

### CODE REVIEW

**Comment by erlend-aasland** [comment]
A couple of questions, @encukou:
- Do I need to ensure GIL is held in `create_callback_context()`, or can I get away with it like it is?
- Ditto for `free_callback_context()`

**Comment by encukou** [comment]
> Do I need to ensure GIL is held in create_callback_context(), or can I get away with it like it is?

There are *very* few funcitons in the C-API that you can call without the GIL held; usually they're the ones needed before a GIL is created. For `PyMem_Malloc`/`PyMem_Free` specifically, there's a  [a big red warning](https://docs.python.org/3.8/c-api/memory.html#memory-interface) to answer your question ;)
`PyMem_Malloc` is [configurable](https://docs.python.org/3.8/c-api/memory.html#customize-memory-allocators); the [default allocator](https://docs.python.org/3.8/c-api/memory.html#the-pymalloc-allocator) relies on the GIL for synchronization.

**Comment by erlend-aasland** [comment]
> There are *very* few funcitons in the C-API that you can call without the GIL held; usually they're the ones needed before a GIL is created. For `PyMem_Malloc`/`PyMem_Free` specifically, there's a  [a big red warning](https://docs.python.org/3.8/c-api/memory.html#memory-interface) to answer your question ;)

Yes, I've seen it ;) But then why does this PR work as it is? (That is, without the GIL)

**Comment by encukou** [comment]
> But then why does this PR work as it is?

I assume it's because you're not using multiple threads.
Even if you were, it might be rare to cause the memory allocator to run in two threads at once. But when it would, it'd crash quite spectacularly. (Or worse, if a malicious actor could cause it to happen.)

**Comment by erlend-aasland** [comment]
> > But then why does this PR work as it is?
> 
> I assume it's because you're not using multiple threads.
> Even if you were, it might be rare to cause the memory allocator to run in two threads at once. But when it would, it'd crash quite spectacularly. (Or worse, if a malicious actor could cause it to happen.)

Ah, right. Thanks!

I've updated the PR. PTAL :)

**Comment by erlend-aasland** [comment]
On hold until #27613 has landed.

File: Modules/_sqlite/connection.c
Comment (encukou) [inline Modules/_sqlite/connection.c]: `Py_NewRef` is also Python API and should be protected by the GIL.
We don't want two threads trying to increment the counter at the time.

File: Modules/_sqlite/connection.c
Comment (encukou) [inline Modules/_sqlite/connection.c:793]: Same here.

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: Got it. Thanks!

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c:793]: PTAL

**Review by encukou — APPROVED** [review]
I found some more questionable uses of the GIL in existing code (see the issue). But let's not block this PR; it is an improvement.

**Pull request merged**
